rp2040-hal/src/gpio.rs

```diff
@@ -1,132 +1,337 @@
 //! General Purpose Input and Output (GPIO)
 //!
-//! TODO: Docs for this whole module
-//!
-//! For example, to turn on the LED on a Pico board:
+//! To access the GPIO pins you must call the `split` method on the IO bank. This will return a
+//! `Parts` struct with access to the individual pins:
 //!
 //! ```rust
+//! use rp2040_hal::prelude::*;
 //! let mut pac = rp2040_pac::Peripherals::take().unwrap();
+//! let pins = pac.IO_BANK0.split(pac.PADS_BANK0, pac.SIO, &mut pac.RESETS);
+//! ```
 //!
-//! let pins = pac.IO_BANK0.split(&mut pac.RESETS);
-//! let mut led_pin = pins.gpio25.into_output(&mut pac.PADS_BANK0, &mut pac.SIO);
-//!  led_pin.set_low().unwrap();
+//! Once you have the GPIO pins struct, you can take individual pins and configure them:
+//!
+//! ```rust
+//! let mut led_pin = pins.gpio25.into_output();
+//! led_pin.set_high().unwrap();
 //! ```
-use core::convert::Infallible;
-use core::marker::PhantomData;
-use embedded_hal::digital::v2::OutputPin;
+//!
+//! Input pins support the following options:
+//! - Pull high, pull low, or floating
+//! - Schmitt trigger
+//!
+//! Output pins support the following options:
+//! - Slew rate (fast or slow)
+//! - Drive strength (2, 4, 8 or 12 mA)
 
-pub struct Floating;
-pub struct Input<MODE> {
-    _mode: PhantomData<MODE>,
-}
+/// Mode marker for an input pin
+pub struct Input;
+/// Mode marker for an output pin
 pub struct Output;
+/// Mode marker for a pin in an unknown state (generally happens at startup)
+pub struct Unknown;
 
-pub trait GpioExt {
+/// This trait adds a method to extract pins from an IO bank and convert them into HAL objects
+pub trait GpioExt<PADS, SIO> {
+    /// The type of struct that will hold the pins once they're converted to HAL objects
     type Parts;
 
+    /// Convert the IO bank into a struct of HAL pin objects
     // TODO: Do we need a marker to check that clocks are up?
-    fn split(self, reset: &mut rp2040_pac::RESETS) -> Self::Parts;
+    fn split(self, pads: PADS, sio: SIO, reset: &mut rp2040_pac::RESETS) -> Self::Parts;
 }
 
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+/// The amount of current that a pin can drive when used as an output
+pub enum OutputDriveStrength {
+    /// 2 mA
+    Ma2,
+    /// 4 mA
+    Ma4,
+    /// 8 mA
+    Ma8,
+    /// 12 mA
+    Ma12,
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+/// The slew rate of a pin when used as an output
+pub enum OutputSlewRate {
+    /// Slew slow
+    Slow,
+    /// Slew fast
+    Fast,
+}
+
+// Magic numbers from the datasheet
+// const FUNCTION_SPI: u8 = 1;
+// const FUNCTION_UART: u8 = 2;
+// const FUNCTION_I2C: u8 = 3;
+// const FUNCTION_PWM: u8 = 4;
 const FUNCTION_SIO: u8 = 5;
+// const FUNCTION_PIO0: u8 = 6;
+// const FUNCTION_PIO1: u8 = 7;
+// const FUNCTION_CLOCK: u8 = 8;
+// const FUNCTION_USB: u8 = 9;
 
 macro_rules! gpio {
```

Comment:
After a good nights sleep I think we can solve this without a macro by using const generics.

Instead of types for `Gpio0`, `Gpio1` etc. there would be just one type  `Gpio<const N: u8>`. This could make the docs easier and less redundant. And might also make it easier for some editors because the macro does not need to be expanded.

But we can also change it in a future iteration. I actually did not yet try const generics in the wild so I do not know what problems might show up.

Reply:
Yes I think for this first round it might be good to stick with what the other HALs are doing, though eliminating the macro would be ideal. The macro makes it way harder to understand what is going on.